yabt/dot.py

```diff
@@ -0,0 +1,61 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2019 Resonai Ltd. All rights reserved
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Creates the dot graph
+~~~~~~~~~~~~~~~~~
+
+:author: Dana Shamir
+"""
+
+
+from networkx.algorithms import descendants
+
+from .caching import get_prebuilt_targets
+from .config import Config
+
+
+TARGETS_COLORS = {'Python': 'red', 'PythonTest': 'pink',
+                  'PythonPackage': 'purple', 'CppLib': 'blue',
+                  'CppProg': 'cornflowerblue', 'CppGTest': 'deepskyblue4',
+                  'AptPackage': 'brown4', 'CustomInstaller': 'brown',
+                  'Proto': 'green'}
+
+
+def write_dot(build_context, conf: Config, out_f):
+    """Write build graph in dot format to `out_f` file-like object."""
+    buildenvs = set(
+        target.buildenv for target in build_context.targets.values()
+        if target.buildenv is not None)
+    buildenv_targets = set(buildenvs)
+    for buildenv in buildenvs:
+        buildenv_targets = buildenv_targets.union(
+            descendants(build_context.target_graph, buildenv))
```

Comment:
this is tricky.
by taking a union over all the transitive deps of all buildenvs you risk missing nodes that *should* be included in the graph, even if buildenvs are filtered.
let me describe with an example:
node A1 uses buildenv B1
node B1 depends on node C1
node A2 uses buildenv B2
node A2 depends on node C1

with this graph, C1 will be filtered out (because it's in the closure of B1), but it's wrong, because C1 is not contained in the closure of B2, so it will be built when building A2.

this feels familiar, and sounds like something I needed to deal with in some other place, so you can try and find it and reuse the same logic (maybe also extract to a common util if it isn't already), assuming that "other place" exists and is really applicable here.

Reply:
Actually I discussed this scenario with @dlehavi-resonai 
We agreed that in this case we still don't want to see these dependencies in the graph.
You want to see them when you are planning third party includes (when turning show_cuildenv_deps flag on), not when you are planning our own code base.
Examples that match this scenario:
3rdparty/misc:glog, 3rdparty/misc:gflags-dev, 3rdparty/apt:ffmpeg, 3rdparty/misc:protobuf, 3rdparty/misc:grpc, 3rdparty/misc:boost, 3rdparty/apt:curl, 3rdparty/apt:zlib, 3rdparty/misc:gflags-dev

Reply:
I don't understand the explanation.

"don't want to see third party deps when focusing on our own codebase" is a valid thing, but is unrelated to the issue I'm describing (maybe it overlaps in the case of the existing codebase, but definitely not generally true).

to do what you want, you can simply add an exclude filter flag, and run with `--exclude-from-dot 3rdparty` (it can be a list-type flag to support multiple simultaneous exclusion, or you can go crazy and take a full regex)

Reply:
@dlehavi-resonai can you answer?
meanwhile @itamaro do you have a guess where I should search your solution for this issue?

Reply:
I think what @itamaro is describing is the simple fact that we cannot do a real automatic separation before separating the build dependencies from the run dependencies. 
What he is than saying that what I suggested you do may be correct for our current code base, but it is wrong for the project which may have more general dependencies.
He is than saying that until we are don with the build / run dependencies separation we can have the exclude policy as a flag.
The problem with this is that the exclusion policy, without taking a transitive closure, and without saying what you do want, may keep a lot of orphans. I may be wrong here - I really don't know. I this problem is fictional, we can go with Itamar's suggestiong (I don't think it's a huge change to what you've already done)

Reply:
I now support this case, also added a test case for it (moved the tests to a different file)